File: server/src/main/java/ch/zxseitz/tbsg/server/games/GameManager.java
package ch.zxseitz.tbsg.server.games;

import ch.zxseitz.tbsg.games.IGame;
import ch.zxseitz.tbsg.games.annotations.Color;
import ch.zxseitz.tbsg.games.annotations.TbsgGame;
import ch.zxseitz.tbsg.games.exceptions.GameException;
import ch.zxseitz.tbsg.server.websocket.Client;
import ch.zxseitz.tbsg.server.websocket.Match;
import ch.zxseitz.tbsg.server.websocket.Protector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider;
import org.springframework.core.type.filter.AnnotationTypeFilter;
import org.springframework.stereotype.Component;

import java.lang.reflect.ParameterizedType;
import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Collectors;

@Component
public class GameManager {
    private static final Logger logger = LoggerFactory.getLogger(GameManager.class);
    private static final String gameBasePackage = "ch.zxseitz.tbsg.games";
//    private static final String gameClassManifest = "Game-Class";

    private final Map<String, GameProxy> proxies;

    // separate method for testing
    public static GameProxy createProxy(Class<?> gameClass) throws GameException {
        var interfaceType = Arrays.stream(gameClass.getGenericInterfaces())
                .map(type -> (ParameterizedType) type)
                .filter(type -> type.getRawType().equals(IGame.class))
                .findFirst().orElseThrow(() -> new GameException("Missing interface IGame", null));
        var actionClass = (Class<?>) interfaceType.getActualTypeArguments()[0];
        var gameAnnotation = gameClass.getAnnotation(TbsgGame.class);
        if (gameAnnotation == null) {
            throw new GameException("Missing @TbsgGame annotation in game class " + gameClass.getSimpleName(), null);
        }
        var gameName = gameAnnotation.name();
        var colors = Arrays.stream(gameAnnotation.colors())
                .collect(Collectors.toMap(Color::value, Color::name));
        if (colors.get(0) != null) {
            throw new GameException("Color " + colors.get(0) + " cannot have value 0.", null);
        }
        return new GameProxy(gameName, (Class<? extends IGame<?>>) gameClass, actionClass, colors);
    }

    public static Match createMatch(IGame<?> game, Map<Integer, Client> clients) {
        return new Match(new Protector<>(game), clients);
    }

    public GameManager() {
        proxies = new HashMap<>(10);
        var provider = new ClassPathScanningCandidateComponentProvider(false);
        provider.addIncludeFilter(new AnnotationTypeFilter(TbsgGame.class));
        var classes = provider.findCandidateComponents(gameBasePackage);
        for (var bean : classes) {
            var className = bean.getBeanClassName();
            logger.info("scanning game class {}", className); //todo debug
            try {
                var gameClass = Class.forName(className);
                var gameProxy = createProxy(gameClass);
                proxies.put(gameProxy.getName(), gameProxy);
                logger.info("registered game {}", gameProxy.getName());
            } catch (Exception e) {
                logger.error(e.getMessage());
            }
        }
    }

    public Set<String> listGameNames() {
        return proxies.keySet();
    }

    public void foreachGame(Consumer<GameProxy> action) {
        proxies.values().forEach(action);
    }
}

File: server/src/main/java/ch/zxseitz/tbsg/server/games/GameManager.kt
package ch.zxseitz.tbsg.server.games

import ch.zxseitz.tbsg.games.IGame
import ch.zxseitz.tbsg.games.annotations.Color
import ch.zxseitz.tbsg.games.annotations.TbsgGame
import ch.zxseitz.tbsg.games.exceptions.GameException
import ch.zxseitz.tbsg.server.websocket.Client
import ch.zxseitz.tbsg.server.websocket.Match
import ch.zxseitz.tbsg.server.websocket.Protector
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider
import org.springframework.core.type.filter.AnnotationTypeFilter
import org.springframework.stereotype.Component

import java.lang.reflect.ParameterizedType
import java.util.*
import java.util.function.Consumer
import java.util.stream.Collectors
import kotlin.collections.HashMap

@Component
class GameManager {
    companion object {
        private val logger: Logger = LoggerFactory.getLogger(GameManager::class.java)
        private const val gameBasePackage: String = "ch.zxseitz.tbsg.games"
//    private val gameClassManifest: String = "Game-Class"

        // separate method for testing
        fun createProxy(gameClass: Class<out IGame<*>>): GameProxy {
            val interfaceType = Arrays.stream(gameClass.genericInterfaces)
                .map{type -> type as ParameterizedType}
                .filter{type -> type.rawType.equals(IGame::class.java)}
                .findFirst().orElseThrow{GameException("Missing interface IGame")}
            val actionClass = interfaceType.actualTypeArguments[0] as Class<*>
            val gameAnnotation = gameClass.getAnnotation(TbsgGame::class.java)
                ?: throw GameException("Missing @TbsgGame annotation in game class ${gameClass.simpleName}")
            val gameName = gameAnnotation.name
            val colors = Arrays.stream(gameAnnotation.colors)
                .collect(Collectors.toMap(Color::value, Color::name))
            if (colors[0] != null) {
                throw GameException("Color ${colors[0]} cannot have value 0.")
            }
            return GameProxy(gameName, colors, gameClass as Class<IGame<Any>>, actionClass)
        }

        fun createMatch(game: IGame<Any>, clients: Map<Int, Client>): Match {
            return Match(Protector(game), clients)
        }
    }

    private var proxies: MutableMap<String, GameProxy> = HashMap()

    init {
        proxies = HashMap(10)
        val provider = ClassPathScanningCandidateComponentProvider(false)
        provider.addIncludeFilter(AnnotationTypeFilter(TbsgGame::class.java))
        val classes = provider.findCandidateComponents(gameBasePackage)
        for (bean in classes) {
            val className = bean.beanClassName
            logger.info("scanning game class $className") //todo debug
            try {
                val gameClass = Class.forName(className) as Class<IGame<Any>>
                val gameProxy = createProxy(gameClass)
                proxies[gameProxy.name] = gameProxy
                logger.info("registered game {}", gameProxy.name)
            } catch (e: Exception) {
                logger.error(e.message)
            }
        }
    }

    val gameNames: Set<String>
        get() =  proxies.keys

    fun foreachGame(action: Consumer<GameProxy>) {
        proxies.values.forEach(action)
    }
}
